[최지수] Step1 리뷰 요청합니다!

jisu1211 opened the pull request:
안녕하세요 리뷰어님!
1단계 인수 테스트 기반 리팩터링 완료하여 리뷰 요청합니다~

1단계 미션은 인수 테스트 주도 개발 미션과 비슷하다보니 리팩터링을 진행하면서
상당 부분 인수 테스트 주도 개발 미션에서 작성했던 코드를 참고하게 되었습니다 ㅎㅎ 
그렇다보니 양이 좀 많아져서 죄송합니다 😢

이번 미션 잘부탁드리겠습니다!
감사합니다 😊

jisu1211 committed:
docs: README.md 작성

jisu1211 committed:
feat: custom exception 추가, exception handler 추가

jisu1211 committed:
feat: 노선 안의 역 목록을 정렬하는 LineStationSorter 추가
test: LineStationSorterTest 추가

jisu1211 committed:
refactor: 구간을 연결하는 기능 수정
- SectionConnectManger : 상행 종점역, 중간역, 하행 종점역에 따라 구간을 연결하는 클래스
- ConnectionType : 구간 연결 타입을 결정하는 클래스

test: ConnectionTypeTest, SectionConnectManagerTest 추가

jisu1211 committed:
refactor: BaseEntity 패키지 이동

jisu1211 committed:
refactor: BaseEntity 패키지 이동

jisu1211 committed:
refactor: 구간 삭제 기능 리팩터링
- SectionDeleteManager : 구간 삭제 전략(상행 종점역, 중간역, 하행 종점역)을 결정하고 실행하는 클래스
- SectionDeleteStrategy : 구간 삭제 전략 인터페이스
- FirstSectionDeleteStrategy : 상행 종점역 구간 삭제
- MiddleSectionDeleteStrategy : 중간역 구간 삭제
- LastSectionDeleteStrategy : 하행 종점역 구간 삭제

test: SectionDeleteManagerTest 추가, SectionDeleteStrategyTest 추가

jisu1211 committed:
refactor: Station 관련 객체 리팩터링
- Station: equalsAndHashCode 수정, 접근제어자 수정, validation 추가, name을 not null로 변경
- StationRepository: Annotation 관련 수정
- StationService: @Transactional 추가
- StationExceptionCode: Station 관련 에러 메시지 보관
- StationController: @RequestMapping 추가

test: Station 관련 테스트 추가(도메인 테스트, 인수 테스트 등)

jisu1211 committed:
refactor: Station 인수 테스트 리팩터링

jisu1211 committed:
refactor: Section 클래스 리팩터링
- Distance 불변 객체 추가
- validateion 추가
- 컬럼 not null 옵션 변경
- 구간 에러 메시지를 보관하는 SectionExceptionCode 추가
- equalsAndHashCode 재정의
- SectionRequest/SectionResponse 추가 및 변경

feat: NumberUtil 추가

test: DistanceTest, SectionTest 추가

jisu1211 committed:
refactor: Line 클래스 리팩터링
- Sections 클래스 추가
- validation 추가
- 구간을 추가/수정/삭제/정렬하는 기능을 도메인 안에서 구현하도록 수정
- equalsAndHashCode 재정의
- 노선의 에러 메시지를 보관하는 LineExceptionCode 추가

test: SectionsTest, LineTest 추가

jisu1211 committed:
refactor: LineService 리팩터링
- 비즈니스 로직을 도메인 안으로 이동, 코드 정리
- Controller 단으로 넘길 데이터들을 dto로 매핑하는 코드 추가

test: LineServiceTest, LineRepositoryTest, SectionRepositoryTest 추가

jisu1211 committed:
refactor: Line, Section 관련 인수 테스트 리팩터링

jisu1211 committed:
refactor: 코드 변경 후 에러 발생하는 부분 수정

geunhwanlee commented on src/main/java/nextstep/subway/common/SubWayExceptionHandler.java:
```diff
@@ -0,0 +1,21 @@
+import org.springframework.web.bind.annotation.ExceptionHandler;
+import org.springframework.web.servlet.mvc.method.annotation.ResponseEntityExceptionHandler;
+
+@ControllerAdvice
+public class SubWayExceptionHandler extends ResponseEntityExceptionHandler {
```
공통 예외처리 👍🏻

geunhwanlee commented on src/main/java/nextstep/subway/line/application/LineService.java:
```diff
@@ -1,182 +1,94 @@
-    private StationService stationService;
+    private final LineRepository lineRepository;
+    private final SectionRepository sectionRepository;
+    private final StationService stationService;
```
서비스에서 서비스를 의존하게되면 순환참조가 발생할 가능성이 있을것같아요!

jisu1211 replied:
넵 안 그래도 이 부분에 고민이 있었는데 수정하도록 하겠습니다!

geunhwanlee commented on src/main/java/nextstep/subway/line/application/LineService.java:
```diff
@@ -1,182 +1,94 @@
+            throw new NotFoundException(LineExceptionCode.NOT_FOUND_BY_ID.getMessage());
+        }
+
+        return optional.get();
```
```suggestion
        return lineRepository.findById(id)
                .orElseThrow(() -> new NotFoundException(LineExceptionCode.NOT_FOUND_BY_ID.getMessage()));
```
이렇게 해도 되겠네요! 😀

geunhwanlee commented on src/main/java/nextstep/subway/line/application/LineService.java:
```diff
@@ -1,182 +1,94 @@
     }
 
-    public List<LineResponse> findLines() {
+    public List<LineResponse> findAllLines() {
```
트랜잭션이 설정되어있지 않네요.
트랜잭션 바깥에서 Lazy Loading 시 예외가 발생합니다!

현재는 OSIV 옵션이 true로 들어가있기 때문에 트랜잭션 바깥에서도 영속성 컨텍스트가 유지되어 예외가 발생하지 않는데요 😀

OSIV 옵션과 무관하게 실행될수 있도록 `@Transactional(readOnly = true)`를 추가하는건 어떨까요?

jisu1211 replied:
` OSIV 옵션`은 생각하지 못했던 부분이네요! 짚어주셔서 감사합니다 😊

geunhwanlee commented on src/main/java/nextstep/subway/line/application/LineService.java:
```diff
@@ -1,182 +1,94 @@
+    public void updateSection(Long lineId, SectionRequest sectionRequest) {
+        Station upStation = stationService.findById(sectionRequest.getUpStationId());
+        Station downStation = stationService.findById(sectionRequest.getDownStationId());
+        List<Section> matchedSections = sectionRepository.findAllByRequestedSection(upStation, downStation);
```
`matchedSections`는 `upStation`과 `downStation`이 둘다 등록되어있거나 아무것도 등록되어있지 않은 경우를 체크하는 용도로만 사용중이네요.

`line`의 `sections`를 통해서 유효성 검사를 하거나
`matchedSections`를 통해서 구간추가를 하는건 어떻게 생각하시나요? 🤔

jisu1211 replied:
근환님이 제안해주신 방향으로 한 번 고민해보겠습니다!

geunhwanlee commented on src/main/java/nextstep/subway/line/domain/ConnectionType.java:
```diff
@@ -0,0 +1,49 @@
+
+import java.util.function.BiFunction;
+
+public enum ConnectionType {
+    FIRST((current, request) -> SectionConnectManager.connectFirstSection(request)),
+    MIDDLE((current, request) -> SectionConnectManager.connectMiddleSection(current, request)),
+    LAST((current, request) -> SectionConnectManager.connectLastSection(request)),
+    NONE((current, request) -> null);
+
+    private BiFunction<Section, Section, Section> expression;
+
+    ConnectionType(BiFunction<Section, Section, Section> expression) {
+        this.expression = expression;
+    }
```
`SectionConnectManager.connectFirstSection`
`SectionConnectManager.connectMiddleSection`
`SectionConnectManager.connectLastSection`
세개 모두 request를 그대로 리턴하고있는데 `BiFunction`을 사용할 필요가 있을까요? 🤔

```suggestion
public enum ConnectionType {
    FIRST((current, request) -> {}),
    MIDDLE(Section::divideSection),
    LAST((current, request) -> {}),
    NONE((current, request) -> {});

    private BiConsumer<Section, Section> expression;

    ConnectionType(BiConsumer<Section, Section> expression) {
        this.expression = expression;
    }

```
`BiConsumer`를 사용하는건 어떨까요? 😀

jisu1211 replied:
제안해주신 코드가 더 깔끔해보이네요! 수정하겠습니다 ㅎㅎ 좋은 의견 감사합니다~

geunhwanlee commented on src/main/java/nextstep/subway/line/domain/Distance.java:
```diff
@@ -0,0 +1,45 @@
+import javax.persistence.Column;
+import javax.persistence.Embeddable;
+
+@Embeddable
+public class Distance {
```
원시값 포장 👍🏻

geunhwanlee commented on src/main/java/nextstep/subway/line/domain/Distance.java:
```diff
@@ -0,0 +1,45 @@
+
+    public Distance plus(int distance) {
+        return new Distance(this.distance + distance);
+    }
```
```suggestion
    public Distance minus(Distance target) {
        if(this.distance <= target.distance) {
            throw new IllegalArgumentException(SectionExceptionCode.INVALID_DISTANCE.getMessage());
        }

        return new Distance(this.distance - target.distance);
    }

    public Distance plus(Distance target) {
        return new Distance(this.distance + target.distance);
    }
```
`Distance`를 정의한김에 `Distance`끼리 연산하게하는건 어떨까요? 😀

geunhwanlee commented on src/main/java/nextstep/subway/line/domain/LineStationSorter.java:
```diff
@@ -0,0 +1,54 @@
+                .filter(section -> section.getUpStation().equals(beforeDownStation))
+                .findFirst();
+
+        return nextSection.isPresent() ? nextSection.get().getDownStation() : null;
```
```suggestion
        return nextSection.map(Section::getDownStation).orElse(null);
```
이렇게 쓸수도 있겠네요! 😀
그리고 개인적으로 null을 리턴하는것보다 Optional을 리턴하는게 낫지 않나 생각해요.

geunhwanlee commented on src/main/java/nextstep/subway/line/domain/LineStationSorter.java:
```diff
@@ -0,0 +1,54 @@
+                .filter(section -> section.getDownStation().equals(beforeUpStation))
+                .findFirst();
+
+        return prevSection.isPresent() ? prevSection.get().getUpStation() : null;
```
```suggestion
        return prevSection.map(Section::getUpStation).orElse(null);
```
😀

geunhwanlee commented on src/main/java/nextstep/subway/line/domain/SectionConnectManager.java:
```diff
@@ -0,0 +1,39 @@
+import static java.util.stream.Collectors.toList;
+
+public class SectionConnectManager {
+    public static void connectAll(Line line, Section request) {
+        List<Section> requests = line.getSections().stream()
+                .map(section -> SectionConnectManager.connect(section, request))
+                .filter(Objects::nonNull)
+                .collect(toList());
+
+        addNewSections(line, requests);
+    }
```
노선에 구간을 하나 추가하는 메서드로 보이는데
`addNewSections(line, requests);`를 보면 구간을 여럿 추가하고있네요!

제가 파악한바에 따르면 기존 구간들 사이즈만큼 request의 리스트가 requests에 담기고 그 리스트를 추가하는것으로 보여요! 😀

jisu1211 replied:
이 부분도 다시 점검해보겠습니다 ㅎㅎ

geunhwanlee commented on src/main/java/nextstep/subway/line/domain/SectionDeleteManager.java:
```diff
@@ -0,0 +1,51 @@
+    }
+
+    private Section toSection(Optional<Section> optSection) {
+        if(optSection.isPresent()) {
+            return optSection.get();
+        }
+
+        return null;
```
```suggestion
        return optSection.orElse(null);

```
이렇게 해도 되겠네요! 😀

geunhwanlee commented on src/main/java/nextstep/subway/line/domain/Section.java:
```diff
@@ -1,45 +1,76 @@
-    public Section(Line line, Station upStation, Station downStation, int distance) {
-        this.line = line;
+    public Section(Station upStation, Station downStation, int distance) {
+        validateStations(upStation, downStation);
+
         this.upStation = upStation;
         this.downStation = downStation;
-        this.distance = distance;
+        this.distance = new Distance(distance);
```
Section은 Line에 종속적일것 같아요!
생성자 파라미터에 Line을 제거하신 이유가 있을까요? 🤔
생성자에 Line을 추가하여 Line을 필수값으로 보장하는게 나을거같아요. 😀

jisu1211 replied:
테스트를 작성하다보니 Line을 생성하면서 Section이 같이 생성되는 경우가 많아서 생성자에 들어가는게 애매해보여서 제거했었는데 다시 추가하도록 하겠습니다!

geunhwanlee commented on src/main/java/nextstep/subway/line/dto/LineResponse.java:
```diff
@@ -27,7 +30,16 @@ public LineResponse(Long id, String name, String color, List<StationResponse> st
+
+    public static List<LineResponse> toLineResponses(List<Line> lines) {
+        return lines.stream()
+                .map(line -> {
+                    return of(line, StationResponse.toStationResponses(line.getSortedStations()));
+                })
```
```suggestion
                .map(line -> of(line, StationResponse.toStationResponses(line.getSortedStations())))
```
이렇게 써볼수도 있겠네요!

또는 `LineResponse of(Line line)` 메서드를 정의하면
```suggestion
                .map(LineResponse::of)
```
이렇게 쓸수도 있겠네요 😀

geunhwanlee commented on src/test/java/nextstep/subway/line/acceptance/SectionAcceptanceTest.java:
```diff
@@ -0,0 +1,151 @@
+
+import java.util.Arrays;
+
+@DisplayName("지하철 구간 관련 기능")
+class SectionAcceptanceTest extends AcceptanceTest {
```
원래 있던 `LineSectionAcceptanceTest`를 삭제한 이유가 있을까요? 🤔
`지하철 노선에 여러개의 역을 순서 상관 없이 등록한다.` 라는 테스트가 사라졌네요. 😭

jisu1211 replied:
삭제하면서 빠진 부분이 있었나보네요 ㅠㅠ 다시 점검하고 추가하도록 하겠습니다!

geunhwanlee reviewed (changes requested):
안녕하세요 인수테스트 기반 TDD 미션을 함께 하게될 이근환입니다. 🙌

1단계 미션 아주 잘 해주셨어요! 👏

몇가지 아쉬운 부분이 보여서 코멘트 드렸어요.

확인해 보시고 재 리뷰요청 주세요. 🙏

그럼 화이팅입니다 🔥

jisu1211 committed:
refactor: LineService/StationService 리팩터링
- @Transaction(readOnly = true) 추가
- orElseThrow를 활용하여 코드 간결하게 수정

jisu1211 committed:
refactor: Distance 리팩터링
- plus, minus 연산시 Distance 객체끼리 연산하도록 수정

jisu1211 committed:
refactor: Section 생성자에 Line 추가
- Line validation 추가

jisu1211 committed:
refactor: 구간 추가 로직 리팩터링
- 요청 구간에 포함된 역을 포함하는 구간 목록을 가져오는 데이터를 구간 추가할 때도 사용하도록 수정
- 값을 리턴할 필요가 없으므로 BiFunction -> BiConsumer로 변경

jisu1211 committed:
refactor: Section 리팩터링
- 생성자에 Line 추가
- validation 추가

jisu1211 committed:
refactor: 코드 가독성 개선

jisu1211 committed:
refactor: Service에서 Service 레이어 참조하지 않도록 수정(순환참조 방지)

jisu1211 committed:
refactor: 코드 가독성 개선

jisu1211 committed:
test: 누락된 인수테스트 추가

jisu1211 commented:
안녕하세요 근환님!
코드가 양이 많았는데 꼼꼼하게 살펴보시고 부족한 점들 피드백 해주셔서 감사합니다 ㅎㅎ
피드백 해주신 부분들 수정 완료했습니다 😊

1. 
> matchedSections는 upStation과 downStation이 둘다 등록되어있거나 아무것도 등록되어있지 않은 경우를 체크하는 용도로만 사용중이네요.
>
> line의 sections를 통해서 유효성 검사를 하거나
> matchedSections를 통해서 구간추가를 하는건 어떻게 생각하시나요? 🤔

2. 
> 노선에 구간을 하나 추가하는 메서드로 보이는데
> addNewSections(line, requests);를 보면 구간을 여럿 추가하고있네요!
>
> 제가 파악한바에 따르면 기존 구간들 사이즈만큼 request의 리스트가 requests에 담기고 그 리스트를 추가하는것으로 보여요! 😀

1번과 2번 피드백 모두 같은 로직에서 주신 부분이라 같이 처리했습니다. 수정한 코드는 아래와 같습니다!
```
public static void connectAll(Line line, Section request, List<Section> matchedSections) {
        Set<Section> requests = matchedSections.stream()
                .map(section -> SectionConnectManager.connect(section, request))
                .filter(Objects::nonNull)
                .collect(toSet());

        addNewSections(line, requests);
}

private static Section connect(Section current, Section request) {
        ConnectionType connectionType = ConnectionType.match(current, request);
        connectionType.connect(current, request);
        return request;
}

private static void addNewSections(Line line, Set<Section> requests) {
        requests.forEach(request -> line.addSection(request));
}
```
원래는 전체 구간을 탐색하면서 처리했던 것을 피드백(1)에 따라 `요청한 상행역 / 하행역을 포함한 기존 구간 리스트인 matchedSections`을 탐색하면서 처리하는 것으로 수정하였습니다.

피드백(2)번에 대해서는 기존에 구간을 전체 탐색했으나 
```
[수정 전 코드]
private static Section connect(Section current, Section request) {
        ConnectionType connectionType = ConnectionType.match(current, request);
        return connectionType.connect(current, request); <<- 여기서 None type일때 null을 리턴했으므로 filter(Objects::nonNull)에 의해 추가되지 않았을 것으로 보입니다! /* NONE((current, request) -> null); */
}
```

현재는 matchedSections으로 탐색 대상을 변경했으니 상행 종점역/하행 종점역 구간일 때는 중복이 발생하지 않고 중간역일 때만 requests가 2개가 될 것 같은데(connect 메서드에서 request를 그대로 리턴하고 있으므로) 그래서 `Set`으로 컬렉션을 변경해봤는데 Set으로 하는게 맞을지 아니면 로직을 좀 더 변경해봐야할지 고민이 되네요! 리뷰 부탁드립니다~

jisu1211 committed:
refactor: Set 컬렉션으로 변경

geunhwanlee commented on src/main/java/nextstep/subway/line/domain/SectionConnectManager.java:
```diff
@@ -0,0 +1,28 @@
+import static java.util.stream.Collectors.toSet;
+
+public class SectionConnectManager {
+    public static void connectAll(Line line, Section request, List<Section> matchedSections) {
+        Set<Section> requests = matchedSections.stream()
+                .map(section -> SectionConnectManager.connect(section, request))
+                .filter(Objects::nonNull)
+                .collect(toSet());
+
+        addNewSections(line, requests);
+    }
```
`matchedSections`는 `request`에 의해 변경되고
`request`는 `line`에 추가되면 될것같은데

```suggestion
    public static void connectAll(Line line, Section request, List<Section> matchedSections) {
        matchedSections.forEach(section -> SectionConnectManager.connect(section, request));

        addNewSection(line, request);
    }

```
이렇게 하는건 어떤가요? 😀

jisu1211 replied:
앗 그러네요..! 굳이 왜 리스트로 받아서 처리하려했는지.. 감사합니다 ㅎㅎㅎ

geunhwanlee commented on src/main/java/nextstep/subway/line/domain/Section.java:
```diff
@@ -1,45 +1,88 @@
-    public Long getId() {
-        return id;
+    public void divideSection(Section request) {
+        this.distance = distance.minus(new Distance(request.getDistance()));
```
```suggestion
        this.distance = distance.minus(request.getDistance());
```
`getDistance`가 `distance`를 그대로 리턴하면 이렇게 쓸수 있을것 같네요.

또는 같은 `Section` 클래스의 값이니
```suggestion
        this.distance = distance.minus(request.distance);
```
이렇게 할수도 있겠네요! 😀

jisu1211 replied:
깔끔하게 개선해주셔서 감사합니다! 많이 배워가네요 ㅎㅎ

geunhwanlee reviewed (approved):
리뷰사항 반영 잘 해주셨어요! 👍🏻

충분히 잘해주셔서 여기서 머지하도록 하겠습니다.

고민하셨던 부분에 대해서 추가적인 코멘트 드렸어요. 😀

그럼 다음 단계도 화이팅하세요 🔥
